helm-org.el

```diff
@@ -281,57 +279,20 @@ nothing to CANDIDATES."
     (helm-exit-and-execute-action
      'helm-org-insert-link-to-heading-at-marker)))
 
-(defun helm-org-refile-within-buffer (buffer)
-  "Refile current heading to another heading in BUFFER.
-Interactively, BUFFER is the current buffer."
-  (interactive (list (current-buffer)))
-  (let* ((org-reverse-note-order t)
-         ;; Checking buffer-base-buffer ensures that we handle
-         ;; indirect buffers, e.g. created with
-         ;; org-tree-to-indirect-buffer
-         (filename (buffer-file-name (or (buffer-base-buffer buffer) buffer)))
-         (target (helm-comp-read "Refile to: "
-                                 (helm-org--get-candidates-in-file filename)))
-         (rfloc (list nil filename nil target)))
-    (org-refile nil nil rfloc)))
-
-(defun helm-org-refile-within-buffer-action (pos)
-  "Action to refile heading at POS to another heading in current buffer.
-POS is an integer or marker, suitable for use with
-`org-with-point-at'."
-  (org-with-point-at pos
-    (helm-org-refile-within-buffer (current-buffer))))
-
-(defun helm-org-run-refile-within-buffer-action ()
-  "Run `helm-org-refile-within-buffer-action'.
-Call this with a keybinding in the Helm buffer."
-  (interactive)
-  (with-helm-alive-p
-    (helm-exit-and-execute-action #'helm-org-refile-within-buffer-action)))
-(put 'helm-org-run-refile-within-buffer-action 'helm-only t)
-
-(defun helm-org-heading-refile (marker)
-  "Refile current heading or marked to MARKER.
-The current heading is the heading where cursor was
-before entering helm session, it will be used unless
-you mark a candidate, in this case helm will go to this marked
-candidate in org buffer and refile this candidate to MARKER.
-NOTE that of course if you have marked more than one candidate,
-all the subsequent candidates will be ignored."
-  (let ((mkd (with-helm-buffer
-               (and helm-marked-candidates
-                    (car (helm-marked-candidates))))))
-    (save-selected-window
-      (when (eq major-mode 'org-agenda-mode)
-        (org-agenda-switch-to))
-      (when mkd (helm-org-goto-marker mkd))
-      (org-cut-subtree)
-      (let ((target-level (with-current-buffer (marker-buffer marker)
-                            (goto-char (marker-position marker))
-                            (org-current-level))))
-        (helm-org-goto-marker marker)
-        (org-end-of-subtree t t)
-        (org-paste-subtree (+ target-level 1))))))
+(defun helm-org--refile-heading-to (marker)
+  "Refile headings to heading at MARKER.
+If multiple candidates are marked in the Helm session, they will
+all be refiled.  If no headings are marked, the selected heading
+will be refiled."
+  (let* ((victims (with-helm-buffer (helm-marked-candidates)))
+         (buffer (marker-buffer marker))
```

Comment:
> Something seems wrong here, if you are now operating on marked candidates, why are you using MARKER, which is a marker refering to the current candidate,

`marker` is the marker of the target heading, i.e. the heading that candidates will be refiled *to*.  It is the `REAL` of the `(DISPLAY . REAL)` pair.  It is only computed once because there is only one target.  It is only used to get the underlying filename of the target heading's buffer.

> I guess you have to compute marker-buffer on each candidates while looping, MARKER will be unused (use _marker).

This is not necessary.  `org-with-point-at` takes a marker as its first argument.  Each `victim` is a marker.  `org-refile` operates with point at an entry, which is what `org-with-point-at` provides.